sklearn/calibration.py

```diff
@@ -202,11 +188,41 @@ def fit(self, X, y, sample_weight=None):
             base_estimator = self.base_estimator
 
         if self.cv == "prefit":
+            # Set `n_features_in_` attribute
+            if isinstance(self.base_estimator, Pipeline):
+                estimator = self.base_estimator[-1]
+            else:
+                estimator = self.base_estimator
+            check_is_fitted(estimator)
+            self.n_features_in_ = estimator.n_features_in_
+            self.classes_ = estimator.classes_
```

Comment:
I can't find the discussing about this.

It seems weird that `CalibratedClassiferCV` takes the `n_features_in_` from the final step. If the first step of the Pipeline was a feature selector, then `CalibratedClassiferCV` would not have correct `n_feature_in_`?

Also, for third party estimators in a pipeline, if they do not have `n_feature_in_` or `classes_` this would fail. I would prefer being slightly more lenient:

```py
with suppress(AttributeError):
    self.n_features_in_ = base_estimator.n_features_in_

with suppress(AttributeError):
	self.classes_ = base_estimator.classes_
```

Reply:
You are right for the `n_features_in_` that it should be the value of the first step and not the last one.

Regarding `classes_`, I would be a bit more conservative. Our `estimator_check` explicitly checks for `classes_` so this is a kind of contract that if you do a scikit-learn classifier, you should have this attribute.

And for `n_features_in_`, we can be lenient in this case because we still don't impose anything yet.

Reply:
WDYT @thomasjpfan ?

Reply:
Yes, you're right. I got mixed up.

Reply:
> Regarding classes_, I would be a bit more conservative. Our estimator_check explicitly checks for classes_ so this is a kind of contract that if you do a scikit-learn classifier, you should have this attribute.

As in don't suppress warning when assigning this attribute?

Reply:
Also out of interest, which of:
```
if hasattr(calib_clf, "n_features_in_"):
    self.n_features_in_ = base_estimator.n_features_in_
```
and
```
with suppress(AttributeError):
    self.n_features_in_ = base_estimator.n_features_in_
```
would be preferred here?

Reply:
I used the first so I don't need to import `suppress` but happy to change

Reply:
I am actually not sure what is more explicit. I am not used to `suppress` but I find it elegant and it should be more pythonic. Go ahead with it.

Reply:
> WDYT @thomasjpfan ?

I agree with the being more strict with `classes_` and be lenient with `n_features_in_`.